### PR TITLE
[WIP] Add HNSW Index

### PR DESCRIPTION
**Pull request by Piachonkin-Alex**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New type of indexes: index for ANN search.

TODO:
- [ ] Documentation
- [x] Detailed description
- [x] Tests

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/MergeTree/CommonANNIndexes.cpp [hunk 1]
```diff
@@ -0,0 +1,518 @@
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTSetQuery.h>
+
+#include <Storages/MergeTree/CommonANNIndexes.h>
+#include <base/logger_useful.h>
```
Comment (FArthur-cmd) [inline src/Storages/MergeTree/CommonANNIndexes.cpp:9]: Wrong path (useless import?)

**Comment by FArthur-cmd** [comment]
@rschu1ze can you watch this pr?

File: src/Storages/MergeTree/MergeTreeIndexHnsw.cpp [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+{
+    auto impl = std::make_unique<HnswWrapper::IndexWrap<float>>("l2");
+    impl->addBatchUnsafe(std::move(data));
+    impl->createIndex(index_params);
+    return std::make_shared<MergeTreeIndexGranuleHnsw>(index_name, index_sample_block, std::move(impl));
+}
+
+void MergeTreeIndexAggregatorHnsw::update(const Block & block, size_t * pos, size_t limit)
+{
+    if (*pos >= block.rows())
```
Comment (rschu1ze) [inline src/Storages/MergeTree/MergeTreeIndexHnsw.cpp:225]: If you use fmt-based exceptions, you wouldn't need `toString()`:

```
throw Exception (
    ErrorCodes::LOGICAL_ERROR,
    "The provided position is not less than the number of block rows. Position: {}, Block rows: {}.",
    *pos, blocks.rows());

File: tests/queries/0_stateless/02354_hnsw.sql [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+FROM 02354_hnsw
+WHERE L2Distance(embedding, [0.0, 0.0, 10.0]) < 1.0
+LIMIT 5;
+
+SELECT *
+FROM 02354_hnsw
+ORDER BY L2Distance(embedding, [0.0, 0.0, 10.0])
+LIMIT 3;
+
+INSERT INTO 02354_hnsw VALUES (1, 1); -- { serverError 53 }
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02354_hnsw.sql:41]: (l 41. doesn't test HNSW-index specific functionality -->  maybe remove?)

File: src/Storages/MergeTree/CommonANNIndexes.cpp [hunk 4]
```diff
@@ -0,0 +1,552 @@
+#include <Storages/MergeTree/CommonANNIndexes.h>
```
[the remark this answers is not in the document]
Comment (FArthur-cmd) [reply src/Storages/MergeTree/CommonANNIndexes.cpp:1]: Yes

File: src/Storages/MergeTree/CommonANNIndexes.h [hunk 5]
```diff
@@ -0,0 +1,228 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (FArthur-cmd) [reply src/Storages/MergeTree/CommonANNIndexes.h:1]: Yes

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 6]
```diff
@@ -1031,3 +1031,58 @@ Examples of working configurations can be found in integration tests directory (
...
+    
+Number of granules in granularity should be large. With greater `GRANULARITY` indexes remember the data structure better. But some indexes can't be built if they don't have enough data, so this granule will always participate in the query. For more information, see the description of indexes.
+
+As the indexes are built only during insertions into table, `INSERT` and `OPTIMIZE` queries are slower than for ordinary table. OAt this stage indexes remember all the information about the given data. ANNIndexes should be used if you have immutable or rarely changed data and many read requests.
+    
+You can create your table with index which uses certain algorithm. Now only indices based on the following algorithms are supported:
+
+##### Index list
+## Hnsw
+- `hnsw([M, efConstruction, graph_degree, maxM, maxM0])` -- Stores and operates with Hierarchical ANN search graph built by [Hnsw algortihm](https://arxiv.org/pdf/1603.09320.pdf). The graph construction algorithm, as well as its structure, very much resembles the Skip list data structure. Short parameters description:
```
[the remark this answers is not in the document]
Comment (FArthur-cmd) [reply docs/en/engines/table-engines/mergetree-family/mergetree.md:1082]: Most of the code was taken from the code of the library itself (there is an implementation for writing to a file, while we need to write to a buffer). But the index is serialized without data, although it needs data to calculate total distances. That's why serialization of the data itself was added. The main unit is the object. It has only 2 constructors (contrib/nmslib/similarity_search/include/object.h)
The second one will call copying, and for all data it is considerably long. I think that vector will not help.

**Pull request closed without merging**